Fix recursive substitutions, add an option to use the XKCD font, and add mouseovers.

piquan opened the pull request:
I had to rewrite much of the substitution code, but tried to keep the original style.  Let me know if there's anything I missed.

This hasn't been heavily tested.

piquan committed:
Add an option to change the font to the XKCD font.

This also will add mouseovers with the original text.

bencoveney commented on substitutions/js/substitutions.js:
```diff
@@ -16,37 +17,88 @@ chrome.runtime.sendMessage("config", function(response) {
+      }
+      var cls = parent.getAttribute("class");
+      if (cls && cls.indexOf("xkcdSubstitutionsExtensionSubbed") != -1) {
         return;
```
Looks like this would catch the nested terms in #28

piquan committed:
Ignore entire subtrees if needed.

We already ignore the immediate children of script, form,
etc. elements, as well as spans that we've inserted ourselves.  This
change makes us skip the entire subtree of such an element.

One important effect is that if another extension changes one of our
spans using a similar algorithm to ours (inserting a new subtree) we
previously could end up reprocessing that node in a future
document_end event.  Now, we prevent that.

piquan commented:
@bencoveney: My original pull request would catch nested terms like #28 describes if we're the only thing mutating the DOM.  However, since we can get called multiple times, if another extension changed our subtree, we might end up reprocessing it.  I've updated the pull request to account for this.

For instance, we substitute "expand" &#8594; "physically expand".  Suppose that another extension classifies words according to their parts of speech.  Now, see how that would play out.

We substitute `expand` &#8594; `<span class="xkcd">physically expand</span>`.  The other extension makes that `<span class="xkcd"><span class="adverb">physically</span> <span class="verb">expand</span></span>`.  Later, we may get another `document_end` event and reprocess the tree, leading to `<span class="xkcd"><span class="adverb">physically</span> <span class="verb"><span class="xkcd">physically expand</span></span></span>`.  You see the problem.

In the code I originally submitted, we only checked the immediate parent to see if it was a span of class `xkcdSubstitutionsExtensionSubbed`.  Clearly, we should behave well in the presence of other DOM-mutating extensions.

I've updated the pull request to prune the entire subtree at the point of `xkcdSubstitutionsExtensionSubbed`.  This also prunes the subtrees of elements on the ignore list, which has elements like `form` etc.  (Again, previously the code only checked the immediate parent, which makes me wonder if `form` should have been in the ignore list in the first place.)

StephenBrown2 committed:
Use small-caps font-variant to fit in better with text on page

piquan committed:
Merge pull request #1 from StephenBrown2/small-caps

Use small-caps font-variant to fit in better with text on page

ravenbuilder934 reviewed (changes requested):
According to the [about page](https://xkcd.com/about/) it should be capitalized as `xkcd`; in multiple instances you have it as `XKCD`
